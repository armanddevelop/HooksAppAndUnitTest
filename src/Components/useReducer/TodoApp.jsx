import React, { useEffect, useReducer } from "react";
import { TodoAdd } from "./Components/TodoAdd";
import { TodoList } from "./Components/TodoList";
import "./styles.css";
import { todoReducer } from "./todoReducer";

const init = () => {
  return JSON.parse(localStorage.getItem("todos")) || [];
};
export const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);
  useEffect(() => {
    localStorage.setItem("todos", JSON.stringify(todos));
  }, [todos]);
  const handleDelete = (todoId) => {
    const action = {
      type: "delete",
      payload: todoId,
    };
    dispatch(action);
  };
  const handleAddToDo = (newTodo) => {
    const action = {
      type: "add",
      payload: newTodo,
    };
    dispatch(action);
  };
  const handleToggle = (todoId) => {
    console.log(todoId);
    const action = {
      type: "toggle",
      payload: todoId,
    };
    dispatch(action);
  };
  return (
    <div>
      <h1>To do App ({todos.length})</h1>
      <hr />
      <div className="row">
        <div className="col-7">
          <TodoList
            todos={todos}
            handleDelete={handleDelete}
            handleToggle={handleToggle}
          />
        </div>
        <div className="col-5">
          <TodoAdd handleAddToDo={handleAddToDo} />
        </div>
      </div>
    </div>
  );
};
