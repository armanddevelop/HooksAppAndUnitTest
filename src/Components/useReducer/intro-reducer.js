const initialState = [
  {
    id: 1,
    todo: "comprar carne",
    done: false,
  },
];

const todoReducer = (state, action) => {
  state = initialState;
  if (action?.type === "add") {
    return [...state, action.payload];
  }
  return state;
};
const newTodo = {
  id: 2,
  todo: "comprar servilletas",
  done: false,
};

const newTodo2 = {
  id: 3,
  todo: "comprar leche",
  done: false,
};

const action = {
  type: "add",
  payload: newTodo2,
};

let todos = todoReducer();
todos = todoReducer(todos, action);

//let todos2 = todoReducer(todos, action);
console.log("esto vale todos ", todos);
