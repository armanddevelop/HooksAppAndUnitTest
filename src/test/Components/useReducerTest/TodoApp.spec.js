import "@testing-library/jest-dom";
import toJson from "enzyme-to-json";
import { shallow, mount } from "enzyme";
import { TodoApp } from "../../../Components/useReducer/TodoApp";
import { demosTodos } from "../../fixtures/demosTodos";
import { act } from "@testing-library/react";

describe("Test for TodoApp component", () => {
  const container = shallow(<TodoApp />);
  Storage.prototype.setItem = jest.fn(() => {});

  it("Snapshot to TodoApp component", () => {
    expect(toJson(container)).toMatchSnapshot();
  });

  it("Should add a to do", () => {
    const container = mount(<TodoApp />);
    act(() => {
      container.find("TodoAdd").prop("handleAddToDo")(demosTodos[0]);
      container.find("TodoAdd").prop("handleAddToDo")(demosTodos[1]);
    });
    expect(container.find("h1").text().trim()).toBe(
      `To do App (${demosTodos.length})`
    );
    expect(localStorage.setItem).toHaveBeenCalledTimes(2);
    //expect(localStorage.setItem).toHaveBeenCalledWith({});
  });

  it("Should delete a todo", () => {
    act(() => {
      container.find("TodoAdd").prop("handleAddToDo")(demosTodos[0]);
      container.find("TodoList").prop("handleDelete")(demosTodos[0].id);
    });

    expect(container.find("h1").text().trim()).toBe(`To do App (${0})`);
  });
});
