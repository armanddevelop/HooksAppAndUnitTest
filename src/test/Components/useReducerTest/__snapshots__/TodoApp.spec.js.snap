// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for TodoApp component Snapshot to TodoApp component 1`] = `
<div>
  <h1>
    To do App (
    0
    )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoAdd
        handleAddToDo={[Function]}
      />
    </div>
  </div>
</div>
`;
